CRM/Gpapi/CaseHandler.php

```diff
@@ -42,6 +42,46 @@ class CRM_Gpapi_CaseHandler {
     9 => 5, // Rejecter => Enquirer
   ];
 
+  public static function addMediumToActivities(int $caseID, int $mediumID) {
+    $caseActivities = civicrm_api3('Activity', 'get', [
+      'case_id' => $caseID,
+    ])['values'];
+
+    foreach ($caseActivities as $activity) {
+      civicrm_api3('Activity', 'create', [
+        'id' => $activity['id'],
+        'medium_id' => $mediumID,
+      ]);
+    }
+  }
+
+  public static function addUTMDataToActivities(int $caseID, array $params) {
+    $utmActivityTypeIDs = civicrm_api3('CustomGroup', 'getvalue', [
+      'name'   => 'utm',
+      'return' => 'extends_entity_column_value',
+    ]);
+
+    $activityTypeOptVals = civicrm_api3('OptionValue', 'get', [
+      'id'     => [ 'IN' => $utmActivityTypeIDs ],
+      'return' => 'value',
+    ])['values'];
+
+    foreach ($utmActivityTypeIDs as $key => $optionValueID) {
+      $utmActivityTypeIDs[$key] = $activityTypeOptVals[$optionValueID]['value'];
+    }
+
+    $relevantCaseActivities = civicrm_api3('Activity', 'get', [
```

Comment:
`$relevantCaseActivities` now contains all activities related to this case. However, we don't actually want to add UTM parameters for *all* case activities, but only the ones we added during this API call (either by creating the case or by applying a timeline). Older activities in this case may have different mediums/UTM values that we don't want to change (or add).

That's where the [snippet](https://redmine.greenpeace.at/issues/12652#note-11) from GP-12652 should help.